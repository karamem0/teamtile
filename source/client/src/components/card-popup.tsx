//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';
import { useDebounce } from 'react-use';

import { SearchIcon } from '@fluentui/react-icons-mdl2';
import {
  ComponentEventHandler,
  Input,
  Popup,
  PopupProps
} from '@fluentui/react-northstar';

import { css } from '@emotion/react';

export interface CardPopupProps<T> {
  values: T[],
  predicate: (filter: string, value: T) => boolean,
  renderer: (values: T[]) => React.ReactNode,
  trigger: React.ReactElement,
  onOpenChange?: ComponentEventHandler<PopupProps>
}

export const CardPopup = <T, >({
  values,
  renderer,
  predicate,
  trigger,
  onOpenChange
}: CardPopupProps<T>): React.ReactElement | null => {

  const [ results, setResults ] = React.useState(values);
  const [ filter, setFilter ] = React.useState<string>('');

  useDebounce(() => {
    if (!filter) {
      setResults(values);
      return;
    }
    setResults(values.filter((value) => predicate(filter, value)));
  }, 500, [
    values,
    filter
  ]);

  const handleChange = React.useCallback((filter: string) => {
    setFilter(filter);
  }, []);

  return (
    <Popup
      content={
        <div
          css={css`
            width: 12rem;
            max-height: 24rem;
            margin: -0.5rem;
            overflow: auto;
          `}>
          <div
            css={css`
              display: flex;
              flex-flow: column;
              gap: 0.25rem;
            `}>
            <Input
              clearable
              fluid
              icon={<SearchIcon />}
              onChange={(_, props) => handleChange(props?.value || '')} />
            {renderer(results)}
          </div>
        </div>
      }
      trigger={trigger}
      onOpenChange={onOpenChange} />
  );

};
