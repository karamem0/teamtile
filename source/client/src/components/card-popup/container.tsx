//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';
import { useDebounce } from 'react-use';

import { InputProps, PopupProps } from '@fluentui/react-northstar';

import { EventHandler } from '../../types/common';

import Presenter from './presenter';

interface CardPopupProps<T> {
  items: T[],
  predicate: (filter: string, item: T) => boolean,
  renderer: (items: T[]) => React.ReactNode,
  trigger: React.ReactElement,
  onOpenChange?: EventHandler<PopupProps>
}

export default function CardPopup <T> ({
  items,
  renderer,
  predicate,
  trigger,
  onOpenChange
}: CardPopupProps<T>): React.ReactElement | null {

  const [ filterItems, setFilterItems ] = React.useState(items);
  const [ filter, setFilter ] = React.useState<string | null>(null);

  useDebounce(() => {
    if (!filter) {
      setFilterItems(items);
    } else {
      setFilterItems(items.filter((item) => predicate(filter, item)));
    }
  }, 500, [
    items,
    filter
  ]);

  const handleInputChange = React.useCallback((_, data?: InputProps & { value: string } | undefined) => {
    setFilter(data?.value || null);
  }, []);

  return (
    <Presenter
      items={filterItems}
      renderer={renderer}
      trigger={trigger}
      onInputChange={handleInputChange}
      onOpenChange={onOpenChange} />
  );

}
