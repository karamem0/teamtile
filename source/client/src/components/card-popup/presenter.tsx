//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { SearchIcon } from '@fluentui/react-icons-mdl2';
import {
  Input,
  InputProps,
  Popup,
  PopupProps
} from '@fluentui/react-northstar';

import { css } from '@emotion/react';

import { EventHandler } from '../../types/common';
import { memo } from '../../utils/memo';

interface CardPopupProps<T> {
  items: T[],
  renderer: (items: T[]) => React.ReactNode,
  trigger: React.ReactElement,
  onInputChange?: EventHandler<InputProps & { value: string }>,
  onOpenChange?: EventHandler<PopupProps>
}

export default memo(function CardPopup <T> ({
  items,
  renderer,
  trigger,
  onInputChange,
  onOpenChange
}: CardPopupProps<T>): React.ReactElement | null {

  return (
    <Popup
      content={
        <div
          css={css`
            width: 12rem;
            max-height: 24rem;
            margin: -0.5rem;
            overflow: auto;
          `}>
          <div
            css={css`
              display: flex;
              flex-flow: column;
              gap: 0.25rem;
            `}>
            <Input
              clearable
              fluid
              icon={<SearchIcon />}
              onChange={(event, data) => onInputChange && onInputChange(event, data)} />
            {renderer(items)}
          </div>
        </div>
      }
      trigger={trigger}
      onOpenChange={(event, data) => onOpenChange && onOpenChange(event, data)} />
  );

});
