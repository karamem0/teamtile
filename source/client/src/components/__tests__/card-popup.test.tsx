//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { render, screen } from '@testing-library/react';

import { CardPopup } from '../card-popup';

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe('CardPopup', () => {

  it('create shapshot', () => {
    const params = {
      values: [],
      renderer: () => <React.Fragment />,
      predicate: jest.fn(),
      trigger: <button role="button" />,
      onOpenChange: jest.fn()
    };
    render(<CardPopup {...params} />);
    screen.getByRole('button').click();
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

});
