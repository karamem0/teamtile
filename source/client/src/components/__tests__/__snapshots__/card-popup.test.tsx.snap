// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPopup create shapshot 1`] = `
.emotion-0 {
  width: 12rem;
  max-height: 24rem;
  margin: -0.5rem;
  overflow: auto;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: column;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  gap: 0.25rem;
}

<body>
  <div>
    <button
      aria-haspopup="dialog"
      data-aa-class="PopupButton"
      role="button"
    />
  </div>
  <div
    class="ui-popup__content ui-animation"
    data-aa-class="Popup__popup"
    data-uses-unhanded-props="true"
    role="complementary"
    style="position: fixed; left: 0px; top: 0px; margin: 0px;"
  >
    <div
      class="ui-popup__content__content"
    >
      <div
        class="emotion-0"
      >
        <div
          class="emotion-1"
        >
          <div
            class="ui-box ui-input"
            data-uses-unhanded-props="true"
          >
            <div
              class="ui-box"
              data-uses-unhanded-props="true"
            >
              <input
                class="ui-box ui-input__input"
                data-aa-class="Input__input"
                data-uses-unhanded-props="true"
                id="ui-input-1"
                type="text"
                value=""
              />
              <div
                class="ui-box ui-input__icon"
                data-uses-unhanded-props="true"
              >
                <span
                  aria-hidden="true"
                  class="root_3aeb045a"
                >
                  <svg
                    class="svg_3aeb045a"
                    focusable="false"
                    viewBox="0 0 2048 2048"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M1344 0q97 0 187 25t168 71 142 110 111 143 71 168 25 187q0 97-25 187t-71 168-110 142-143 111-168 71-187 25q-125 0-239-42t-211-121l-785 784q-19 19-45 19t-45-19-19-45q0-26 19-45l784-785q-79-96-121-210t-42-240q0-97 25-187t71-168 110-142T989 96t168-71 187-25zm0 1280q119 0 224-45t183-124 123-183 46-224q0-119-45-224t-124-183-183-123-224-46q-119 0-224 45T937 297 814 480t-46 224q0 119 45 224t124 183 183 123 224 46z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
